src/main/java/org/apache/tools/bzip2/CBZip2InputStream.java

```diff
@@ -29,957 +29,918 @@
-			setupNoRandPartA();
-		}
-	}
-	
-	private static final class Data extends Object {
-		
-		// (with blockSize 900k)
-		final boolean[] inUse = new boolean[256]; //      256 byte
-		
-		final byte[] seqToUnseq = new byte[256]; //      256 byte
-		final byte[] selector = new byte[MAX_SELECTORS]; //    18002 byte
-		final byte[] selectorMtf = new byte[MAX_SELECTORS]; //    18002 byte
-		
-		/**
-		 * Freq table collected to save a pass over the data during
-		 * decompression.
-		 */
-		final int[] unzftab = new int[256]; //     1024 byte
-		
-		final int[][] limit = new int[N_GROUPS][MAX_ALPHA_SIZE]; //     6192 byte
-		final int[][] base = new int[N_GROUPS][MAX_ALPHA_SIZE]; //     6192 byte
-		final int[][] perm = new int[N_GROUPS][MAX_ALPHA_SIZE]; //     6192 byte
-		final int[] minLens = new int[N_GROUPS]; //       24 byte
-		
-		final int[] cftab = new int[257]; //     1028 byte
-		final char[] getAndMoveToFrontDecode_yy = new char[256]; //      512 byte
-		final char[][] temp_charArray2d = new char[N_GROUPS][MAX_ALPHA_SIZE]; //     3096 byte
-		final byte[] recvDecodingTables_pos = new byte[N_GROUPS]; //        6 byte
-		//---------------
-		//    60798 byte
-		
-		int[] tt; //  3600000 byte
-		byte[] ll8; //   900000 byte
-		
-		//---------------
-		//  4560782 byte
-		//===============
-		
-		Data(int blockSize100k) {
-			super();
-			
-			this.ll8 = new byte[blockSize100k * BZip2Constants.baseBlockSize];
-		}
-		
-		/**
-		 * Initializes the {@link #tt} array.
-		 *
-		 * This method is called when the required length of the array is known.
-		 * I don't initialize it at construction time to avoid unneccessary
-		 * memory allocation when compressing small files.
-		 */
-		final int[] initTT(int length) {
-			int[] ttShadow = this.tt;
-			
-			// tt.length should always be >= length, but theoretically
-			// it can happen, if the compressor mixed small and large
-			// blocks.  Normally only the last block will be smaller
-			// than others.
-			if ((ttShadow == null) || (ttShadow.length < length)) {
-				this.tt = ttShadow = new int[length];
-			}
-			
-			return ttShadow;
-		}
-		
-	}
+                if (nextSym <= 16) {
+                    for (int j = nextSym - 1; j > 0; ) {
+                        yy[j] = yy[--j];
+                    }
+                } else {
+                    System.arraycopy(yy, 0, yy, 1, nextSym - 1);
+                }
+
+                yy[0] = tmp;
+
+                if (groupPos == 0) {
+                    groupPos = G_SIZE - 1;
+                    zt = selector[++groupNo] & 0xff;
+                    base_zt = base[zt];
+                    limit_zt = limit[zt];
+                    perm_zt = perm[zt];
+                    minLens_zt = minLens[zt];
+                } else {
+                    groupPos--;
+                }
+
+                int zn = minLens_zt;
+
+                // Inlined:
+                // int zvec = bsR(zn);
+                while (bsLiveShadow < zn) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor the code in order to not assign to this loop counter from within the loop body. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://titan.maxgamer.org:9000/coding_rules#rule_key=squid%3AForLoopCounterChangedCheck)